ui-modules/blueprint-composer/app/views/main/main.controller.js

```diff
@@ -110,8 +117,37 @@ export function MainController($scope, $element, $log, $state, $stateParams, brB
                 $log.error('Cannot save layers preferences: ' + ex.message);
             }
         }
+    };
+
+    // Relationship arcs and labels.
+    let isRelationshipArcsLayerActive = true;
+    let isRelationshipLabelsLayerActive = true;
+
+    // Re-apply filters when selected filters changed or graph is changed.
+    $scope.$watch('vm.layers', () => {
+        applyFilters();
+
+        let relationshipArcsLayer = vm.layers.find(l => l.id === PATHS_SELECTOR_ID);
+        let relationshipLabelsLayer = vm.layers.find(l => l.id === LABELS_SELECTOR_ID);
+
+        // Turning on labels turns on arcs.
+        if (relationshipLabelsLayer.active && !isRelationshipLabelsLayerActive) {
+            relationshipArcsLayer.active = true;
+        }
+
+        // Turning off arcs turns off labels.
+        if (!relationshipArcsLayer.active && isRelationshipArcsLayerActive) {
+            relationshipLabelsLayer.active = false;
+        }
+
+        // Update cached state for arc and labels.
+        isRelationshipArcsLayerActive = relationshipArcsLayer.active;
+        isRelationshipLabelsLayerActive = relationshipLabelsLayer.active;
+
     }, true);
 
+    $scope.$on('layers.filter', () => applyFilters());
```

Comment:
Why do we need to apply the filter when the graph is redrawn? Sounds like the logic is backward. The entrypoint is the click on the filters, isn't it?

Reply:
We need to redraw it for cases when filter is changed first and graph is changed after. Filter change in Layers applies changes styling changes to all existing elements on the canvas. Re-draw event is to re-apply those styling effects for graph changes.

Reply:
Aha, I see @algairim. That is a bit counter-intuitive but probably the best solution for now (and I wouldnt want to make you change even more stuff now :P)